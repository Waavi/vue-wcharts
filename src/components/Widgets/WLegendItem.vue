<template>
    <a
        class="WLegendItem"
        :title="text"
        :class="{ active }"
        :style="styles"
        @click.prevent="$emit('onClick', { text, index })"
        @mouseenter="$emit('onMouseenter', { text, index })"
        @mouseleave="$emit('onMouseleave', { text, index })"
    >
        <slot
            name="bullet"
            :index="index"
            :text="text"
            :color="colorCmp"
        >
            <WBullet
                :borderColor="colorCmp"
                color="white"
            />
        </slot>
        <slot :text="text">
            <span class="Text">{{ text }}</span>
        </slot>
    </a>
</template>

<script>
import VueTypes from 'vue-types'
import themeMixin from '../../mixins/theme'
import WBullet from './WBullet.vue'

export default {
    name: 'WLegendItem',
    components: {
        WBullet,
    },
    mixins: [themeMixin],
    props: {
        index: VueTypes.oneOfType([Number, String]),
        text: VueTypes.string,
        active: VueTypes.bool.def(false),
        styles: VueTypes.object,
        textStyles: VueTypes.object,
        color: VueTypes.string,
    },
    computed: {
        // Get color value. Color prop, Color theme by index or color by defualt
        colorCmp () {
            if (this.color) return this.color
            if (this.index) return this.colors[this.index]
            return this.colors[0]
        },
    },
}
</script>

<style scoped lang="scss">
.WLegendItem {
    display: flex;
    align-items: center;
    font-size: 12px;
    opacity: .5;
    transition: opacity .3s ease;
    cursor: pointer;

    &.active {
        opacity: 1;
    }
}
</style>
