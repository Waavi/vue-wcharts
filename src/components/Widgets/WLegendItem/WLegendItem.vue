<template>
    <a
        :title="text"
        :style="stylesCmp"
        v-on="legentItemListeners"
    >
        <slot
            name="bullet"
            :index="index"
            :text="text"
            :color="bulletColor"
        >
            <WBullet :styles="{ ...bulletStylesCmp, borderColor: bulletColor }" />
        </slot>
        <slot :text="text">
            <span>{{ text }}</span>
        </slot>
    </a>
</template>

<script>
import VueTypes from 'vue-types'
import merge from 'lodash.merge'
import themeMixin from '../../../mixins/theme'
import animationMixin from '../../../mixins/animation'
import WBullet from '../WBullet/WBullet.vue'

export default {
    name: 'WLegendItem',
    components: {
        WBullet,
    },
    mixins: [themeMixin, animationMixin],
    props: {
        index: VueTypes.oneOfType([Number, String]),
        text: VueTypes.string,
        active: VueTypes.bool.def(false),
        color: VueTypes.string,
        styles: VueTypes.object,
        disabledStyles: VueTypes.object,
        bulletStyles: VueTypes.object,
    },
    computed: {
        // Event Listeners
        legentItemListeners () {
            return merge({}, this.$listeners, {
                click: (event) => {
                    const { text, index } = this
                    this.$emit('onClick', { text, index })
                },
                mouseenter: (event) => {
                    const { text, index } = this
                    this.$emit('onMouseenter', { text, index })
                },
                mouseleave: () => {
                    const { text, index } = this
                    this.$emit('onMouseleave', { text, index })
                },
            })
        },
        // Styles
        stylesCmp () {
            return {
                transition: this.transition,
                ...this.themeStyles.styles,
                ...this.styles,
                ...(this.active ? {} : { ...this.themeStyles.disabledStyles, ...this.disabledStyles }),
                ...(this.trigger === 'click' ? { cursor: 'pointer' } : {}),
            }
        },
        bulletStylesCmp () {
            return {
                ...this.themeStyles.bulletStyles,
                ...this.bulletStyles,
            }
        },
        // Get color value. Color prop, Color theme by index or color by default
        bulletColor () {
            if (this.color) return this.color
            if (this.index) return this.colors[this.index]
            return this.colors[0]
        },
    },
}
</script>
