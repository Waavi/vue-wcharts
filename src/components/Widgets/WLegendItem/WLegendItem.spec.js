import { shallowMount, mount } from '@vue/test-utils'
import WLegendItem from './WLegendItem.vue'

describe('Widgets/WLegendItem', () => {
    const propsData = {
        text: 'test',
        trigger: 'click',
        styles: {
            color: '#000',
            bulletStyles: {
                background: '#eee',
            },
        },
    }

    const defaultConfig = {
        propsData,
    }

    xit(`Should be render correctly`, () => {
        const wrapper = shallowMount(WLegendItem, defaultConfig)
        expect(wrapper.html()).toMatchSnapshot()
    })

    xit('It emits the handleClick event', () => {
        const wrapper = mount(WLegendItem, defaultConfig)
        wrapper.find('a').trigger('click')
        expect(wrapper.emitted('onClick')).toHaveLength(1)
    })

    xit('It emits the handleMouseEnter event', () => {
        const wrapper = mount(WLegendItem, defaultConfig)
        wrapper.find('a').trigger('mouseenter')
        expect(wrapper.emitted('onMouseenter')).toHaveLength(1)
    })

    xit('It emits the handleMouseleave event', () => {
        const wrapper = mount(WLegendItem, defaultConfig)
        wrapper.find('a').trigger('mouseleave')
        expect(wrapper.emitted('onMouseleave')).toHaveLength(1)
    })
})
