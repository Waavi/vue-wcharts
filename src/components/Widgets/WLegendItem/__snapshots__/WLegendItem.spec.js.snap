// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Widgets/WLegendItem Should be render correctly 1`] = `
<a title="" class="WLegendItem" style="transition: all 1s ease; color: rgb(0, 0, 0);">
  <wbullet-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" styles="[object Object]"></wbullet-stub> <span class="Text"></span>
</a>
`;
